test/test_optimization_problem.py

```diff
@@ -0,0 +1,256 @@
+    composite: TestVar = dataclasses.field(default_factory=TestVar)
+    parameter: StorageType = default_storage_field(Parameter)
+
+    def __post_init__(self):
+        self.parameter = np.zeros(3)
+
+
+def test_opti_solver_with_parameters():
+    problem = OptimizationProblem()
+    initial_guess = TestVarAndPar()
+    var = problem.generate_optimization_objects(input_structure=TestVarAndPar())
+    np.random.seed(123)
+    a = 10.0 * np.random.rand(3) + 0.01
+    b = 20.0 * np.random.rand(3) - 10.0
+    c = 20.0 * np.random.rand(3) - 10.0
+
+    initial_guess.parameter = c
+
+    problem.add_expression(
+        mode=ExpressionType.minimize,
+        expression=(
+            a[k] * cs.power(var.composite.variable[k], 2)
+            + b[k] * var.composite.variable[k]
+            for k in range(0, 3)
+        ),
+    )
+
+    problem.add_expression(
+        mode=ExpressionType.subject_to,
+        expression=(  # noqa
+            var.composite.variable[k] >= var.parameter[k] for k in range(3)
+        ),
+    )
+
+    problem.solver().set_initial_guess(initial_guess=initial_guess)
+
+    output, cost_value = problem.solver().solve()
+
+    expected_x = np.zeros(3)
+    expected_cost = 0
+    for i in range(3):
+        expected = -b[i] / (2 * a[i])
+        expected_x[i] = expected if expected >= c[i] else c[i]
+        expected_cost += (
+            -b[i] ** 2 / (4 * a[i])
+            if expected >= c[i]
+            else a[i] * (c[i] ** 2) + b[i] * c[i]
+        )
+
+    assert output.composite.variable == pytest.approx(expected_x)
+    assert cost_value == pytest.approx(expected_cost)
+    assert output.parameter == pytest.approx(c)
+
+    assert problem.solver().get_solution().composite.variable == pytest.approx(
+        expected_x
+    )
+    assert problem.solver().get_cost_value() == pytest.approx(expected_cost)
+
+
+def test_opti_solver_with_parameters_and_lists():
+    problem = OptimizationProblem()
+    initial_guess = []
+    for _ in range(3):
+        initial_guess.append(TestVarAndPar())
+
+    var = problem.generate_optimization_objects(input_structure=initial_guess)
+    np.random.seed(123)
+
+    a = []
+    b = []
+    c = []
+
+    for j in range(len(initial_guess)):
+        a.append(10.0 * np.random.rand(3) + 0.01)
+        b.append(20.0 * np.random.rand(3) - 10.0)
+        c.append(20.0 * np.random.rand(3) - 10.0)
+        initial_guess[j].parameter = c[j]
+
+    problem.add_expression(
+        mode=ExpressionType.minimize,
+        expression=(
+            a[j][k] * cs.power(var[j].composite.variable[k], 2)
+            + b[j][k] * var[j].composite.variable[k]
+            for j in range(len(initial_guess))
+            for k in range(0, 3)
+        ),
+    )
+
+    problem.add_expression(
+        mode=ExpressionType.subject_to,
+        expression=(  # noqa
+            var[j].composite.variable[k] >= c[j][k]
+            for j in range(len(initial_guess))
+            for k in range(3)
+        ),
+    )
```

Comment:
I am mostly looked at how user code looks, and I was a bit confused by the `add_expression` method. First of all, I was a bit confused by `add_expression` meant at a first glance, perhaps separated methods for cost such as `add_constraint` and `add_cost` make more sense. 
More specifically, it is not clear to me what happens when calls `problem.add_expression(mode=ExpressionType.minimize` . Is this substituting the existing cost, or being added to it? If the second case is true, probably a good name for this function could be `add_to_cost`. If we have functions with just one argument, we could also drop the named argument `expression=`, and increse the information density reducing the noise, i.e. :

~~~py
    problem.add_to_cost(
        (
            a[j][k] * cs.power(var[j].composite.variable[k], 2)
            + b[j][k] * var[j].composite.variable[k]
            for j in range(len(initial_guess))
            for k in range(0, 3)
        ),
    )

    problem.add_constraint(
        (  # noqa
            var[j].composite.variable[k] >= c[j][k]
            for j in range(len(initial_guess))
            for k in range(3)
        ),
    )
~~~

Reply:
For example, Drake uses `add_cost` and `add_constraint`, see https://github.com/RobotLocomotion/drake/blob/v1.14.0/tutorials/mathematical_program.ipynb .

Reply:
The idea was to allow defining an expression, and then deciding at runtime what you want to do with it. For the time being the options are ``minimize``, ``subject_to``, and ``skip``. This is an idea I borrowed from the `blf`` ik where each task has a given priority that can be set from a configuration file. The idea is to easily allow moving a cost portion to constraints and vice-versa.

> Is this substituting the existing cost, or being added to it? 

At the moment every call adds a new cost (or constraint). There is no coming back.

Regarding the named argument, this is something of ``python``, it is not mandatory, but usually suggested.

Anyhow, if you consider it confusing, I can also add two specific methods ``add_constraint`` and ``add_cost`` that can easily coexist with the ``add_expression`` versions for people that can't decide until the last moment

Reply:
Option B. That's why it is called ``add_expression`` and not ``set_expression``

Reply:
> Anyhow, if you consider it confusing, I can also add two specific methods add_constraint and add_cost that can easily coexist with the add_expression versions for people that can't decide until the last moment

I think this may be convenient. Just for me to understand, can you make an example of a case in which you first add an expression as `minimize`, and then you convert it to `subject_to`? It is not clear to me how this can work as there is no expression identifier.

Reply:
I am assuming you wanted to set ``ExpressionType.minimize``, because otherwise you are adding two constraints

Reply:
> I am assuming you wanted to set `ExpressionType.minimize`, because otherwise you are adding two constraints

Yes, sorry, I fixed it now.

Reply:
> > Anyhow, if you consider it confusing, I can also add two specific methods add_constraint and add_cost that can easily coexist with the add_expression versions for people that can't decide until the last moment
> 
> I think this may be convenient. Just for me to understand, can you make an example of a case in which you first add an expression as `minimize`, and then you convert it to `subject_to`? It is not clear to me how this can work as there is no expression identifier.

You might set the mode from a configuration file, thus enforcing or relaxing a task without editing the code.

Reply:
> > > Anyhow, if you consider it confusing, I can also add two specific methods add_constraint and add_cost that can easily coexist with the add_expression versions for people that can't decide until the last moment
> > 
> > 
> > I think this may be convenient. Just for me to understand, can you make an example of a case in which you first add an expression as `minimize`, and then you convert it to `subject_to`? It is not clear to me how this can work as there is no expression identifier.
> 
> You might set the mode from a configuration file, thus enforcing or relaxing a task without editing the code.

Ack, so basically you save an if. My intuition is that by adding also `add_cost` and `add_constraint` on top of `add_expression` you get the best of both worlds.

Reply:
Done in https://github.com/ami-iit/hippopt/pull/4/commits/109859cd371d860eb18993ae3bedd249d5e4ca2c